File: PInvoke/NativeTypeIterator.vb
' Copyright (c) Microsoft Corporation.  All rights reserved.
Imports System.Collections.Generic

''' <summary>
''' Parent Child symbol
''' </summary>
''' <remarks></remarks>
Public Class NativeSymbolRelationship
    Private _parent As NativeSymbol
    Private _symbol As NativeSymbol

    Public ReadOnly Property Parent() As NativeSymbol
        Get
            Return _parent
        End Get
    End Property

    Public ReadOnly Property Symbol() As NativeSymbol
        Get
            Return _symbol
        End Get
    End Property

    Public Sub New(ByVal parent As NativeSymbol, ByVal symbol As NativeSymbol)
        _parent = parent
        _symbol = symbol
    End Sub
End Class

''' <summary>
''' Used to perform various iterations on NativeType
''' </summary>
''' <remarks></remarks>
Public Class NativeSymbolIterator

    ''' <summary>
    ''' Find all of the reachable NativeType instances from this one
    ''' </summary>
    ''' <param name="nt"></param>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public Function FindAllReachableDefinedTypes(ByVal nt As NativeType) As IEnumerable(Of NativeDefinedType)

        Dim list As List(Of NativeSymbol) = FindAllNativeSymbols(nt)
        Dim retList As New List(Of NativeDefinedType)
        For Each ns As NativeSymbol In list
            If ns.Category = NativeSymbolCategory.Defined Then
                retList.Add(DirectCast(ns, NativeDefinedType))
            End If
        Next

        Return retList
    End Function

    Public Function FindAllNativeSymbols(ByVal enumerable As IEnumerable(Of NativeSymbol)) As List(Of NativeSymbol)
        Dim found As List(Of NativeSymbolRelationship) = FindAllReachableChildrenImpl(enumerable)
        Dim list As New List(Of NativeSymbol)
        For Each rel As NativeSymbolRelationship In found
            list.Add(rel.Symbol)
        Next

        Return list
    End Function

    Public Function FindAllNativeSymbols(ByVal ns As NativeSymbol) As List(Of NativeSymbol)
        If ns Is Nothing Then : Throw New ArgumentNullException("ns") : End If

        Dim list As New List(Of NativeSymbol)
        list.Add(ns)
        Return FindAllNativeSymbols(list)
    End Function

    Public Function FindAllNativeSymbolRelationships(ByVal ns As NativeSymbol) As List(Of NativeSymbolRelationship)
        Dim list As New List(Of NativeSymbol)
        list.Add(ns)
        Return FindAllNativeSymbolRelationships(list)
    End Function

    Public Function FindAllNativeSymbolRelationships(ByVal enumerable As IEnumerable(Of NativeSymbol)) As List(Of NativeSymbolRelationship)
        Return FindAllReachableChildrenImpl(enumerable)
    End Function

    ''' <summary>
    ''' Process all of the reachable children and return the list of found items
    ''' </summary>
    ''' <param name="originalToProcess"></param>
    ''' <remarks></remarks>
    Private Function FindAllReachableChildrenImpl(ByVal originalToProcess As IEnumerable(Of NativeSymbol)) As List(Of NativeSymbolRelationship)

        Dim found As New List(Of NativeSymbolRelationship)
        Dim map As New Dictionary(Of NativeSymbol, Boolean)
        Dim toVisit As New Queue(Of NativeSymbol)(originalToProcess)

        ' First add in all of the original symbols with no parents
        For Each orig As NativeSymbol In originalToProcess
            found.Add(New NativeSymbolRelationship(Nothing, orig))
        Next

        While toVisit.Count > 0
            Dim cur As NativeSymbol = toVisit.Dequeue()
            If map.ContainsKey(cur) Then
                Continue While
            End If

            map(cur) = True
            For Each child As NativeSymbol In cur.GetChildren()
                found.Add(New NativeSymbolRelationship(cur, child))
                toVisit.Enqueue(child)
            Next
        End While

        Return found
    End Function


End Class

File: PInvoke.Core/NativeTypeIterator.cs
// Copyright (c) Microsoft Corporation.  All rights reserved.
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;

namespace PInvoke
{

    /// <summary>
    /// Parent Child symbol
    /// </summary>
    /// <remarks></remarks>
    public class NativeSymbolRelationship
    {
        public NativeSymbol Parent { get; }
        public NativeSymbol Symbol { get; }

        public NativeSymbolRelationship(NativeSymbol parent, NativeSymbol symbol)
        {
            Parent = parent;
            Symbol = symbol;
        }

        public override string ToString() => $"{Parent} -> {Symbol}";
    }

    /// <summary>
    /// Used to perform various iterations on NativeType
    /// </summary>
    /// <remarks></remarks>
    public class NativeSymbolIterator
    {

        /// <summary>
        /// Find all of the reachable NativeType instances from this one
        /// </summary>
        /// <param name="nt"></param>
        /// <returns></returns>
        /// <remarks></remarks>
        public IEnumerable<NativeDefinedType> FindAllReachableDefinedTypes(NativeType nt)
        {

            List<NativeSymbol> list = FindAllNativeSymbols(nt);
            List<NativeDefinedType> retList = new List<NativeDefinedType>();
            foreach (NativeSymbol ns in list)
            {
                if (ns.Category == NativeSymbolCategory.Defined)
                {
                    retList.Add((NativeDefinedType)ns);
                }
            }

            return retList;
        }

        public List<NativeSymbol> FindAllNativeSymbols(IEnumerable<NativeSymbol> enumerable)
        {
            List<NativeSymbolRelationship> found = FindAllReachableChildrenImpl(enumerable);
            List<NativeSymbol> list = new List<NativeSymbol>();
            foreach (NativeSymbolRelationship rel in found)
            {
                list.Add(rel.Symbol);
            }

            return list;
        }

        public List<NativeSymbol> FindAllNativeSymbols(NativeSymbol ns)
        {
            if (ns == null)
            {
                throw new ArgumentNullException("ns");
            }

            List<NativeSymbol> list = new List<NativeSymbol>();
            list.Add(ns);
            return FindAllNativeSymbols(list);
        }

        public List<NativeSymbolRelationship> FindAllNativeSymbolRelationships(NativeSymbol ns)
        {
            List<NativeSymbol> list = new List<NativeSymbol>();
            list.Add(ns);
            return FindAllNativeSymbolRelationships(list);
        }

        public List<NativeSymbolRelationship> FindAllNativeSymbolRelationships(IEnumerable<NativeSymbol> enumerable)
        {
            return FindAllReachableChildrenImpl(enumerable);
        }

        /// <summary>
        /// Process all of the reachable children and return the list of found items
        /// </summary>
        /// <param name="originalToProcess"></param>
        /// <remarks></remarks>
        private List<NativeSymbolRelationship> FindAllReachableChildrenImpl(IEnumerable<NativeSymbol> originalToProcess)
        {

            List<NativeSymbolRelationship> found = new List<NativeSymbolRelationship>();
            Dictionary<NativeSymbol, bool> map = new Dictionary<NativeSymbol, bool>();
            Queue<NativeSymbol> toVisit = new Queue<NativeSymbol>(originalToProcess);

            // First add in all of the original symbols with no parents
            foreach (NativeSymbol orig in originalToProcess)
            {
                found.Add(new NativeSymbolRelationship(null, orig));
            }

            while (toVisit.Count > 0)
            {
                NativeSymbol cur = toVisit.Dequeue();
                if (map.ContainsKey(cur))
                {
                    continue;
                }

                map[cur] = true;
                foreach (NativeSymbol child in cur.GetChildren())
                {
                    found.Add(new NativeSymbolRelationship(cur, child));
                    toVisit.Enqueue(child);
                }
            }

            return found;
        }
    }
}
